### DISCUSSION TITLE
Implementation of user-defined annotation converters

### DISCUSSION THREAD

**Discussion by Sigmanificient** [EpikCord/EpikCord.py#191] [Ideas]
The idea behind this feature is to use python typing ability is order to automatically convert a argument passed as string to a custom type with a behavior defined by the user

```py
@client.prefix_command()
async def ping(message, thing: CustomConverter):
    assert isinstance(thing, CustomConverter)  # True
```

This discussion is meant to forward idea on the implementation of that feature

**Comment by TheUntraceable** [comment]
This would only be applicable for prefix command btw. Unless for the Application Command we can have stuff like `CustomUser`, I don't know.

**Reply by EmmmaTech** [reply]
What do you mean? This could easily be extended to Slash Commands.
If you set the type of the option to `str` then you can do your own special parsing

**Comment by Sigmanificient** [comment]
A implementation using a register binding a Type to a command.

There is multiple pros:
- Function based so small user implementation
- Single hook store with ability to override bas hook 
- Hot-swappable with function (register_arg_type, remove_arg_type, etc)
-  Keep a free structure that should be maintainable & scalable


```py
from EpikCord import User, Client

client = Client(...)


def user_converter(value):
    return CustomThing(value)

client.register_type(user_converter, CustomThing)
# client.register_type(hook=user_converter, annotation=CustomThing)


@client.prefix_command()
async def ping(message, user: CustomThing):
    ...
```
